master/internal/checkpoint_gc.go

```diff
@@ -116,6 +116,15 @@ func (t *checkpointGCTask) Receive(ctx *actor.Context) error {
 		}, t.db, t.rm, t.taskLogger)
 
 		t.allocation, _ = ctx.ActorOf(t.allocationID, allocation)
+
+		t.Base.TaskContainerDefaults, err = t.rm.TaskContainerDefaults(
+			ctx,
+			rp,
+			model.TaskContainerDefaultsConfig{})
```

Comment:
you need to pass `config.GetMasterConfig().TaskContainerDefaults` here, otherwise it will not pick up the master config ones.. this api needs a refactor to not require this.

Reply:
But it's still picking up the task container defaults from the resource pools and overrides from "master.yaml", as per the testing section in the EE pull request https://github.com/determined-ai/determined-ee/pull/827

Notice that "--nodelist=node009" and "--exclude=node008" from the "rig_test_resource" pool, which is the "default_aux_resource_pool", as well as "--mem-per-cpu=200" from the "cpuQ" partition overrides are getting added.

```
Custom slurm arguments: [--nodelist=node009 --exclude=node008 --mem-per-cpu=200] 
```

The resource pools in "master.yaml" contain the following:

```
        resource_pools:
        - pool_name: rig_test_pool
          description: My aux pool
          provider:
            type: hpc
            partition: cpuQ
          task_container_defaults:
            slurm:
              sbatch_args:
                - --nodelist=node009
                - --exclude=node008
        - pool_name: custom_defq_GPU
          description: My GPU pool
          task_container_defaults:
            slurm:
              sbatch_args:
                - -xnode009
          provider:
            type: hpc
            partition: defq_GPU
```

The partition overrides in "master.yaml" contain the following:

```
        resource_manager:
          master_host: casablanca-login
          master_port: 8085
          host: localhost
          port: 18085
          protocol: http
          container_run_type:
          security:
             tls:
                skip_verify: true
          type: slurm
          auth_file: /Users/rcorujo/.casablanca_login.token
          tres_supported: true
          gres_supported: true
          rendezvous_network_interface:
          default_aux_resource_pool: rig_test_pool
          default_compute_resource_pool: custom_defq_GPU
          slot_type:
          partition_overrides:
             cpuQ:
                task_container_defaults:
                   slurm:
                      sbatch_args:
                         - --mem-per-cpu=200
             defq_GPU:
                task_container_defaults:
                   slurm:
                      sbatch_args:
                         - --exclusive
                         - --mem-per-cpu=100
```

Delete the experiment.

```
(base) rcorujo@C02FJ0T8MD6Q cifar10_pytorch % det experiment delete --yes 8516        
Deletion of experiment 8516 is in progress
```

The Slurm arguments from the "default_aux_resource_pool" and "cpuQ" partition overrides are all consumed.

```
master: DEBU[2023-05-01T11:00:20-04:00] Custom slurm arguments: [--nodelist=node009 --exclude=node008 --mem-per-cpu=200]  actor-local-addr=launcherRM actor-system=master allocation-id=5983f1bb-7e3e-465c-b1d8-a489d0fe134d.1 go-type=dispatcherResourceManager

...

master: DEBU[2023-05-01T11:00:39-04:00] Delete job successful 1cf6ae91-af9c-4168-9df4-fcd7cd0b8b30  actor-local-addr=launcherRM actor-system=master go-type=dispatcherResourceManager
```

Reply:
sorry, i was unclear

Reply:
Ok, I added it to the root of master config, as you mentioned.

```
        port: 8081
        db:
          host: localhost
          port: 5431
          password: postgres
          user: postgres
          name: determined
        checkpoint_storage:
          type: shared_fs
          host_path: /mnt/lustre/foundation_engineering/determined-cp
        log:
          level: debug
        cache:
          cache_dir: /tmp
        task_container_defaults:
          bind_mounts:
            - host_path: /etc/hosts
              container_path: /etc/hosts
          slurm:
            sbatch_args:
              - --mincpus=1
```

I see that it's now getting included with your suggested change of "config.GetMasterConfig().TaskContainerDefaults".

```
master: DEBU[2023-05-01T15:27:03-04:00] Custom slurm arguments: [--nodelist=node009 --exclude=node008 --mem-per-cpu=200 --mincpus=1]  actor-local-addr=launcherRM actor-system=master allocation-id=ea2654c7-2b2d-4905-b827-64c6e43577c7.1 go-type=dispatcherResourceManager
```

However, what if there are Slurm arguments in the root "task_container_defaults" that are GPU specific and may cause problems with the GC task that only runs on CPU nodes?

Reply:
| However, what if there are Slurm arguments in the root "task_container_defaults" that are GPU specific and may cause problems with the GC task that only runs on CPU nodes?

I guess it will break :-) -- though they shouldn't put things in task_container_defaults that are not the default for all jobs.  There are already 3 levels of inheritance -- so it seems we have every use case covered somehow.